File: front/src/lib/api/article.ts
import { type HttpClient, ApiError } from "@/lib/http"

export interface ArticleItem {
  bgPicture: string
  category: string
  comment: number
  created_at: string
  id: string
  is_new: boolean
  tags: string[]
  title: string
  state?: "published" | "draft" | "takeoff"
}

interface ArticleResponse {
  list: ArticleItem[]
  total: number
}

interface ArticleBackendResponse {
  code: number
  message: string
  data: ArticleResponse
}

interface saveArticlePayload {
  title: string
  content: string
  tags: string[]
  category: string
  bgPicture: string
}

interface saveArticleResponse {
  code: number
  message: string
}

interface articleDetailBackendResponse {
  code: number
  message: string
  data: ArticleItem & { content: string }
}

export interface ArticleApi {
  getArticles: () => Promise<ArticleResponse>
  getAllArticles: () => Promise<ArticleResponse>
  saveArticle: (payload: saveArticlePayload) => Promise<saveArticleResponse>
  getArticleDetail: (id: string) => Promise<ArticleItem & { content: string }>
}

const PUBLISHED_ARTICLE_API_URL = "/blogs/articleList"
const ALL_ARTICLE_API_URL = "/dashboard/allArticle"
const SAVE_ARTICLE_API_URL = "/save"
const ARTICLE_DETAIL_API_URL = "/articleDetail"

export const createArticleApi = (client: HttpClient): ArticleApi => {
  return {
    async getArticles() {
      try {
        const response = await client.get<ArticleBackendResponse>(PUBLISHED_ARTICLE_API_URL)
        console.log("获取文章列表成功:", response)
        if (!response.data || (response.code !== 0 && response.code !== 200)) {
          throw new ApiError({
            status: 500,
            message: response.message || "Failed to load articles",
            code: String(response.code),
            details: response,
          })
        }
        return response.data
      } catch (error) {
        throw new ApiError({
          status: 500,
          message: "Failed to load articles",
          code: String(500),
          details: error,
        })
      }
    },
    async getAllArticles() {
      try {
        const response = await client.get<ArticleBackendResponse>(ALL_ARTICLE_API_URL)
        if (!response.data || (response.code !== 0 && response.code !== 200)) {
          throw new ApiError({
            status: 500,
            message: response.message || "Failed to load articles",
            code: String(response.code),
            details: response,
          })
        }
        return response.data
      } catch (error) {
        throw new ApiError({
          status: 500,
          message: "Failed to load articles",
          code: String(500),
          details: error,
        })
      }
    },
    async saveArticle(payload: saveArticlePayload) {
      try {
        const response = await client.post<saveArticleResponse>(SAVE_ARTICLE_API_URL, payload)
        if (response.code !== 0 && response.code !== 200) {
          throw new ApiError({
            status: 500,
            message: response.message || "Failed to save article",
            code: String(response.code),
            details: response,
          })
        }
        return response
      } catch (error) {
        throw new ApiError({
          status: 500,
          message: "Failed to load articles",
          code: String(500),
          details: error,
        })
      }
    },
    async getArticleDetail(id) {
      try {
        const response = await client.get<articleDetailBackendResponse>(`${ARTICLE_DETAIL_API_URL}?id=${id}`)
        if (!response.data || (response.code !== 0 && response.code !== 200)) {
          throw new ApiError({
            status: 500,
            message: response.message || "Failed to load article detail",
            code: String(response.code),
            details: response,
          })
        }
        return response.data
      } catch (error) {
        throw new ApiError({
          status: 500,
          message: "Failed to load article detail",
          code: String(500),
          details: error,
        })
      }
    },
  }
}

File: backend/models/article.py
from sqlalchemy import Column, String, Text, Boolean, Integer, DateTime, JSON
from database import Base
import datetime

class Article(Base):
    __tablename__ = "articles"

    id = Column(String(36), primary_key=True, index=True)
    title = Column(String(255), nullable=False)
    content = Column(Text, nullable=False)
    created_at = Column(DateTime, default=datetime.datetime.utcnow)
    updated_at = Column(DateTime, default=datetime.datetime.utcnow, onupdate=datetime.datetime.utcnow)
    tags = Column(JSON, nullable=True)  # 存储为 JSON 数组
    category = Column(String(100), nullable=True)
    is_new = Column(Boolean, default=True)
    bgPicture = Column(String(500), nullable=True)
    state = Column(String(20), default="draft")  # published, draft, takeoff
    comment = Column(Integer, default=0)
    views = Column(Integer, default=0)  # 文章查看次数
    author = Column(String(100), nullable=True)
    location = Column(String(100), nullable=True)  # 文章发布者地理位置

    def to_dict(self):
        return {
            "id": self.id,
            "title": self.title,
            "content": self.content,
            "created_at": self.created_at.strftime("%Y-%m-%d") if self.created_at else None,
            "updated_at": self.updated_at.strftime("%Y-%m-%d %H:%M:%S") if self.updated_at else None,
            "tags": self.tags,
            "category": self.category,
            "is_new": self.is_new,
            "bgPicture": self.bgPicture,
            "state": self.state,
            "comment": self.comment,
            "views": self.views,
            "author": self.author,
            "location": self.location
        }
